### PR TITLE
expression: add json into builtin if function.

### PR DESCRIPTION
**Pull request by hicqu**

### CODE REVIEW

**Comment by shenli** [comment]
LGTM

**Comment by zz-jason** [comment]
should builtin function `NULLIF` consider json input as well ?

**Comment by XuHuaiyu** [comment]
Nullif has been rewritten to `IF`. @zz-jason

**Comment by zz-jason** [comment]
@XuHuaiyu ok

**Comment by zz-jason** [comment]
@hicqu any update ?

**Comment by hicqu** [comment]
@zz-jason PTAL, thanks!

File: util/types/field_type.go [hunk 1]
```diff
@@ -346,6 +347,9 @@ func DefaultTypeForValue(value interface{}, tp *FieldType) {
 		SetBinChsClnFlag(tp)
 	case json.JSON:
 		tp.Tp = mysql.TypeJSON
+		tp.Decimal = UnspecifiedLength
```
Comment (zz-jason) [inline util/types/field_type.go]: 1. why set `Decimal` with `UnspecifiedLength ` ?
in mysql:
```sql
create table t(a json);
MySQL > select a from t;
Field   1:  `a`
Catalog:    `def`
Database:   `test`
Table:      `t`
Org_table:  `t`
Type:       JSON
Collation:  binary (63)
Length:     4294967295
Max_length: 0
Decimals:   0
Flags:      BLOB BINARY
```
2. should we set `Flen`  here ?

File: expression/builtin_control.go [hunk 2]
```diff
@@ -327,6 +330,7 @@ func (c *ifFunctionClass) getFunction(args []Expression, ctx context.Context) (s
 		return nil, errors.Trace(err)
 	}
 	retTp := c.inferType(args[1].GetType(), args[2].GetType())
+
 	evalTps := fieldTp2EvalTp(retTp)
```
Comment (zz-jason) [inline expression/builtin_control.go]: `fieldTp2EvalTp` is not tacking `tpJSON` into consideration, will we do it in the future ?

Comment (hicqu) [inline expression/builtin_control.go]: @XuHuaiyu , Could you help me on this?

**Comment by zz-jason** [comment]
@hicqu plz resolve conflicts, @XuHuaiyu PTAL

**Comment by zz-jason** [comment]
to #4080

File: expression/builtin_control.go (hunk 2 above)
Comment (XuHuaiyu) [reply expression/builtin_control.go]: sorry for late replay
you can add JSON in fieldTp2EvalTp as the following code:
``` sql
	case types.ClassString:
		switch tp.Tp {
		case mysql.TypeDate, mysql.TypeDatetime:
			return tpDatetime
		case mysql.TypeTimestamp:
			return tpTimestamp
		case mysql.TypeDuration:
			return tpDuration
		}
                case mysql.TypeJSON:
                       return tpJSON
	}
```

**Comment by XuHuaiyu** [comment]
please resolve the conflicts.

**Comment by hicqu** [comment]
@XuHuaiyu  PTAL, thanks.

**Comment by XuHuaiyu** [comment]
LGTM
please resolve the conflicts.

**Comment by zz-jason** [comment]
@hicqu ci failed due to:
```
expression/builtin_control.go:420: undefined: tipb.ScalarFuncSig_IfJson
expression/builtin_control.go:600: undefined: tipb.ScalarFuncSig_IfNullJson
```

**Comment by hicqu** [comment]
Got it, thanks. I will raise a PR in `tipb`.

**Comment by zz-jason** [comment]
you can remain this pr and merge it after https://github.com/pingcap/tipb/pull/55 being merged.

**Pull request merged**